src/java/org/apache/cassandra/index/sai/disk/v1/postings/PostingsWriter.java

```diff
@@ -0,0 +1,314 @@
+    private final long[] deltaBuffer;
+    private final LongArrayList blockOffsets = new LongArrayList();
+    private final LongArrayList blockMaxIDs = new LongArrayList();
+    private final RAMIndexOutput inMemoryOutput = new RAMIndexOutput("blockOffsets");
+
+    private final long startOffset;
+
+    private int bufferUpto;
+    private long lastSegmentRowId;
+    private long maxDelta;
+    private long totalPostings;
+
+    public PostingsWriter(IndexDescriptor indexDescriptor, IndexContext indexContext, boolean segmented) throws IOException
+    {
+        this(indexDescriptor, indexContext, BLOCK_SIZE, segmented);
+    }
+
+    public PostingsWriter(IndexOutput dataOutput) throws IOException
+    {
+        this(dataOutput, BLOCK_SIZE);
+    }
+
+    @VisibleForTesting
+    PostingsWriter(IndexDescriptor indexDescriptor, IndexContext indexContext, int blockSize, boolean segmented) throws IOException
+    {
+        this(indexDescriptor.openPerIndexOutput(IndexComponent.POSTING_LISTS, indexContext, true, segmented), blockSize);
+    }
+
+    private PostingsWriter(IndexOutput dataOutput, int blockSize) throws IOException
+    {
+        this.blockSize = blockSize;
+        this.dataOutput = dataOutput;
+        startOffset = dataOutput.getFilePointer();
+        deltaBuffer = new long[blockSize];
+        SAICodecUtils.writeHeader(dataOutput);
+    }
+
+    /**
+     * @return current file pointer
+     */
+    public long getFilePointer()
+    {
+        return dataOutput.getFilePointer();
+    }
+
+    /**
+     * @return file pointer where index structure begins
+     */
+    public long getStartOffset()
+    {
+        return startOffset;
+    }
+
+    /**
+     * write footer to the postings
+     */
+    public void complete() throws IOException
+    {
+        SAICodecUtils.writeFooter(dataOutput);
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        dataOutput.close();
+    }
+
+    /**
+     * Encodes, compresses and flushes given posting list to disk.
+     *
+     * @param postings posting list to write to disk
+     *
+     * @return file offset to the summary block of this posting list
+     */
+    public long write(PostingList postings) throws IOException
+    {
+        checkArgument(postings != null, "Expected non-null posting list.");
+        checkArgument(postings.size() > 0, "Expected non-empty posting list.");
+
+        resetBlockCounters();
+        blockOffsets.clear();
+        blockMaxIDs.clear();
+
+        long segmentRowId;
+        // When postings list are merged, we don't know exact size, just an upper bound.
+        // We need to count how many postings we added to the block ourselves.
+        int size = 0;
+        while ((segmentRowId = postings.nextPosting()) != PostingList.END_OF_STREAM)
+        {
+            writePosting(segmentRowId);
+            size++;
+            totalPostings++;
+        }
+        if (size == 0)
+            return -1;
```

Comment:
nit: Any tests that hit this path?

Reply:
I have changed this to `assert size > 0 : "No postings were written"` because it should never be possible to get here with `size == 0`. As such, there are no tests that could have hit the path.